### PR TITLE
refactor of ai_fn

### PR DESCRIPTION
**Pull request by aaazzam**
- ai_fn resembles ai_choice: there's a FunctionSystem and FunctionUser prompt which is subclassable. 
- ai_fn now accepts a 'system' kwarg: if the passed system argument is a Prompt it will overwrite the default system prompt, if the passed argument is a string it will add it as a prefix to the default system prompt. 
- This let's users get complete finegrained control over ai_fn. 

- ai_fn is also now a subclass of Function, a general 'pydantic-like' object that gives access to methods like 'source_code', etc.

### CODE REVIEW

File: src/marvin/components/ai_function.py [hunk 1]
```diff
@@ -1,105 +1,85 @@
...
-        `FormatResponse` requires keyword arguments, so pass your response under
-        the `data` parameter for validation.
-        {% endif %}
-        """),
-    prompt_library.User(content="""
-        {% if input_binds %} 
+        {{ description }}
+        {% endif %}\
+    """
+    description: str = ""
```
Comment (jlowin) [inline src/marvin/components/ai_function.py:43]: Are these attributes extraneous? It looks like they are templated in from the actual AIFn class. I *think* the `Message`-based prompts automatically attempt to pass their own attributes as render kwargs so may be better to remove them to avoid accidental overwriting

Comment (jlowin) [reply src/marvin/components/ai_function.py:43]: Ah I see that AIFunction reimplements e.g. source_code as a property

**Comment by jlowin** [comment]
This feels a little overwrought to me, there are so many different paths that open up based on implicit typing (e.g. pass a string, get a prefix to the main prompt, pass a prompt and replace the prompt wholesale is a very hard-to-reason-about split, because everywhere else system prompt objects are merged and therefore have the "prefix" behavior). The use of @property instead of validating the inputs and storing them natively as Pydantic fields also looks nonstandard to my eye. 

It seems like instead of introducing so many ways to modify the AIFunction model, the same could be achieved more cleanly by:

- having `AIFunction` subclass `Function` and take an optional `system` kwarg that's either a string or a prompt and is added to the default prompt render chain (yielding "prefix" behavior)
	- also give it a `.prompts()` method that returns the prompts it actually uses at runtime

- letting users create their own subclass and override just the `.prompts()` method to have full control of the prompt

In other words I dont think theres a need to do so much work to let users wholesale overwrite the prompt here as long as they can create base classes that can do so. In general, I think it is unlikely that anyone would overwrite a default prompt entirely unless they are 1) getting an entirely new behavior (which is a good time for a distinct subclass) or 2) relying on automatic templating of new attributes that they added to the model (which would require a distinct subclass)

**Comment by jlowin** [comment]
Closing following #455, can reopen when ready
